src/basic/path-util.c

```diff
@@ -703,47 +721,64 @@ int find_executable_full(const char *name, const char *root, bool use_path_envva
                 if (r == 0)
                         break;
 
-                if (!path_is_absolute(element))
-                        continue;
+                if (find_executable_from_path(name, root, element, &last_error, ret_filename, ret_fd, fd))
+                        return 0;
+        }
 
-                if (!path_extend(&element, name))
-                        return -ENOMEM;
+        return last_error;
+}
 
-                if (root) {
-                        char *path_name;
+bool find_executable_from_path(
+        const char *name,
+        const char *root,
+        const char *element,
+        int *last_error_addr,
+        char **ret_filename,
+        int *ret_fd,
+        int fd) {
+        int r;
+        char *dup_element = strdup(element);
```

Comment:
Need to check for null. Also needs to be declared with `_cleanup_free_` or `dup_element` could leak.